README.md
# CostOf.Life

[![Build](https://github.com/noandrea/costoflife-rs/actions/workflows/rust.yml/badge.svg)](https://github.com/noandrea/costoflife-rs/actions/workflows/rust.yml)
[![Coverage Status](https://coveralls.io/repos/github/noandrea/costoflife-rs/badge.svg?branch=master)](https://coveralls.io/github/noandrea/costoflife-rs?branch=master)
[![dependency status](https://deps.rs/repo/github/noandrea/costoflife-rs/status.svg)](https://deps.rs/repo/github/noandrea/costoflife-rs)
[![Crates.io](https://img.shields.io/crates/v/costoflife)](https://crates.io/crates/costoflife)

This is the CostOf.Life library and command line client.

For a more general description about the project check website: [thecostof.life](https://thecostof.life) 


## Installation 

install it with `cargo install costoflife`


## Examples 

To add a single transaction use the command `costoflife add`:

```
|> costoflife add Netflix 7.99€ 100320 1m12x .movies .covid
```

See it in action:

[![asciicast](https://asciinema.org/a/382419.svg)](https://asciinema.org/a/382419)


## Transactions

The library interpret transactions and calculate for each transaction its _per diem_ amount. For a longer explanation check the [CostOf.life](https://thecostof.life) website.

A transaction as the properties:
- Name: a descriptive name
- Amount: the amount in monetary currency
- Lifetime: a duration that the transaction applies to
- Start date: the start date since when the lifetime should be computed
- Tags: for grouping transactions

As a simple interface the library interprets strings into transactions, the format of the string is shown in the examples and the details are listed below.
### Parsing rules 

The library tokenize the input string and looks for the patterns listed below. Anything that cannot be recognized as a pattern it will set as the title of the transaction. The title is **required**

#### Amount 

The monetary value of the transaction, **required**:

```EBNF
Amount ::= Natural ( '.' Digit Digit? )? "€" 

Natural ::= NaturalDigit Digit*
NaturalDigit ::= #'[1-9]'
Digit ::= "0" | NaturalDigit 
```

Currently the only available currency is `€`

Examples:
- `10€`
- `10000.99€`


#### Lifetime

The duration of transaction, optional, defaults to `1d`.
    
```EBNF
Lifetime ::= Duration Repeat?

Duration ::= Natural TimeUnit
Repeat ::= Natural "x"
TimeUnit ::= "d" | "w" | "m" | "y"
``` 

where the `TimeUnit` is:
- `d` days
- `w` weeks
- `m` months
- `y` years 

Examples:
- `1m12x` => one month for 12 times, for example for monthly expenses like monthly subscriptions (Netflix, etc)
- `12m` => twelve months for 1 time, same as `1y`
- `1w52x` => one week 52 times, for example weekly groceries expenses for all the year

> 💡 the number of repeats they influence the total amount of the transaction: `10€ 1m12x` will result of a transaction of total amount of `120€` while `12m1x` will result in a single transaction of `10€` over 12 months 


#### Start date

The transaction start date, optional, defaults to the current date, it uses the little endian format (day, month, year).

```EBNF
Date ::= Day Month Year

Month ::= "1" #'[0-2]' | "0" NaturalDigit
Day ::= '0' #'[1-9]' | #'[1-2]' Digit | '3' #'[0-1]'
Year ::= Digit Digit

Natural ::= NaturalDigit Digit*
NaturalDigit ::= #'[1-9]'
Digit ::= "0" | NaturalDigit 
```

Examples:
- `030521` => March the 3rd, 2021
- `312122` => December the 31st, 2022

#### Tags

To label transactions, optional. For convenience it uses the hashtag format.

```EBNF
HashTag ::=  ('#' | '.')  Word

EOL ::= '\r'? '\n' 
Word ::= AlphaNum+ [ (' ' | '\t')+ | EOL ]
AlphaNum  ::= #'[A-Za-z0-9_-]'
```

Examples:
- `#lifestile` 
- `.whatever`

## Appendix

Here the full grammar

```EBNF

Tx ::= ( Amount | Lifetime | StartDate | HashTag ) ( SEP ( Amount | Lifetime | StartDate | HashTag )  )+ EOL

EOL ::= '\r'? '\n' 
SEP ::= (' ' | '\t')+

Lifetime ::= Duration Repeat?

Duration ::= Natural TimeUnit
Repeat ::= Natural "x"
TimeUnit ::= "d" | "w" | "m" | "y"

StartDate ::= Day Month Year

Month ::= "1" #'[0-2]' | "0" NaturalDigit
Day ::= '0' #'[1-9]' | #'[1-2]' Digit | '3' #'[0-1]'
Year ::= Digit Digit

Amount ::= Natural ( '.' Digit Digit? )? "€" 

Natural ::= NaturalDigit Digit*
NaturalDigit ::= #'[1-9]'
Digit ::= "0" | NaturalDigit 
Int ::= "+" | "-" Digit+

HashTag ::=  ('#' | '.')  Word

Word ::= AlphaNum+
AlphaNum  ::= #'[A-Za-z0-9_-]'
```

---
Made by [adgb](https://adgb.me)


## Techincal Setup

1. Install rust 
2. Install cargo-edit: `cargo install cargo-edit`. Cargo edit is used to manage dependencies. 
3. Install cargo-bump: `cargo install cargo-bump`. Cargo bump is used to bump release versions.


CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="0.3.2"></a>
## [0.3.2] - 2022-05-03
### Chore
- improve ergonomics of get_progress


<a name="0.3.1"></a>
## [0.3.1] - 2022-05-02
### Build
- **ci:** add github action setup

### Chore
- updated dependencies
- rename master branch to main
- minor library changes
- update deps
- reorganize structure

### Feat
- print the totals for search results
- **cli:** add search command to search for tx

### Fix
- use local time to get the tody date


<a name="0.3.0"></a>
## [0.3.0] - 2021-01-08
### Feat
- **cli:** add summary for tags
- **lib:** implement the cost_of_life calculation


<a name="0.2.4"></a>
## [0.2.4] - 2021-01-03
### Chore
- **cli:** minor tweaks to the cli interface

### Doc
- better documentation

### Fix
- regexp for tags


<a name="0.2.3"></a>
## [0.2.3] - 2021-01-02
### Doc
- update readme

### Tests
- add wasm tests


<a name="0.2.2"></a>
## [0.2.2] - 2021-01-02
### Chore
- **lib:** improve error management and test coverage

### Feat
- **lib:** tx tags are returned in alphabetical order

### Fix
- **lib:** end_date should be the last inclusive date
- **lib:** magic number for 100% should be 1.0
- **lib:** date calculation off of one day

### Test
- **cli:** add tests for datastore


<a name="0.2.1"></a>
## [0.2.1] - 2020-12-28
### Fix
- **cli:** honour the "--on" parameter
- **cli:** print the correct app name and version

### Test
- **lib:** add more tests for the lib part


<a name="0.2.0"></a>
## [0.2.0] - 2020-12-27
### Chore
- fix clippy warnings

### Feat
- add summary command
- **wasm:** simplify return types


<a name="0.1.2"></a>
## [0.1.2] - 2020-12-07
### Feat
- print the cost of life from the cli

### Fix
- bumping version to fix a wasm-pack issue


<a name="0.1.1"></a>
## [0.1.1] - 2020-12-07
### Build
- add makefile

### Chore
- refactor code and add wasm support

### Feat
- persist data
- add basic command line
- add support for months
- **lang:** accept . (dot) as a tag prefix
- **lib:** add progess calculation


<a name="0.1.0"></a>
## 0.1.0 - 2020-11-30

[Unreleased]: https://github.com/noandrea/costoflife-rs/compare/0.3.2...HEAD
[0.3.2]: https://github.com/noandrea/costoflife-rs/compare/0.3.1...0.3.2
[0.3.1]: https://github.com/noandrea/costoflife-rs/compare/0.3.0...0.3.1
[0.3.0]: https://github.com/noandrea/costoflife-rs/compare/0.2.4...0.3.0
[0.2.4]: https://github.com/noandrea/costoflife-rs/compare/0.2.3...0.2.4
[0.2.3]: https://github.com/noandrea/costoflife-rs/compare/0.2.2...0.2.3
[0.2.2]: https://github.com/noandrea/costoflife-rs/compare/0.2.1...0.2.2
[0.2.1]: https://github.com/noandrea/costoflife-rs/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/noandrea/costoflife-rs/compare/0.1.2...0.2.0
[0.1.2]: https://github.com/noandrea/costoflife-rs/compare/0.1.1...0.1.2
[0.1.1]: https://github.com/noandrea/costoflife-rs/compare/0.1.0...0.1.1


src/utils.rs
use bigdecimal::BigDecimal;
use chrono::{DateTime, FixedOffset, Local, NaiveDate};
use std::str::FromStr;

pub fn parse_amount(s: &str) -> Option<BigDecimal> {
    BigDecimal::from_str(s).ok()
}

/// Returns the current date
pub fn today() -> NaiveDate {
    Local::today().naive_utc()
}

/// Returns the datetime with the local timezone
pub fn now_local() -> DateTime<FixedOffset> {
    DateTime::from(Local::now())
}

/// Builds a date from day/month/year numeric
///
/// # Examples
///
/// ```
/// use costoflife;
///
/// let d = costoflife::date(1, 1, 2001); // 2001-01-01
///
/// ```
pub fn date(d: u32, m: u32, y: i32) -> NaiveDate {
    NaiveDate::from_ymd(y, m, d)
}

/// Parse a date from string, it recognizes the formats
///
/// - dd/mm/yyyy
/// - dd.mm.yyyy
/// - ddmmyy
/// - dd.mm.yy
/// - dd/mm/yy
///
pub fn date_from_str(s: &str) -> Option<NaiveDate> {
    let formats = vec!["%d%m%y", "%d.%m.%y", "%d/%m/%y", "%d/%m/%Y", "%d.%m.%Y"];
    // check all the formats
    for f in formats {
        let r = NaiveDate::parse_from_str(s, f);
        if r.is_ok() {
            return r.ok();
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parsers() {
        // parse date
        let r = date_from_str("27/12/2020");
        assert_eq!(r.unwrap(), date(27, 12, 2020));
        // invalid date
        let r = date_from_str("30/02/2020");
        assert_eq!(r, None);
        // invalid format
        let r = date_from_str("30/02/20");
        assert_eq!(r, None);
        // dd.mm.yy
        let r = date_from_str("30.01.20");
        assert_eq!(r.unwrap(), date(30, 1, 2020));
        // dd/mm/yy
        let r = date_from_str("30/01/20");
        assert_eq!(r.unwrap(), date(30, 1, 2020));
        // dd.mm.yyyy
        let r = date_from_str("30/01/2020");
        assert_eq!(r.unwrap(), date(30, 1, 2020));
    }
}


src/ledger.rs
use ::costoflife::{self, TxRecord};
use bigdecimal::{BigDecimal, ToPrimitive};
use chrono::NaiveDate;
use simsearch::SimSearch;
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufRead, LineWriter, Write};
use std::path::Path;
use std::str::FromStr;

/// A simple datastore that can persist data on file
///
pub struct DataStore {
    data: HashMap<blake3::Hash, TxRecord>,
    index: SimSearch<blake3::Hash>,
}
impl DataStore {
    /// Initialize an empty datastore
    ///
    pub fn new() -> DataStore {
        DataStore {
            data: HashMap::new(),
            index: SimSearch::new(),
        }
    }
    /// Load the datastore with the records found
    /// at log_file path
    pub fn load(&mut self, log_file: &Path) -> Result<(), std::io::Error> {
        // read path
        if let Ok(lines) = DataStore::read_lines(log_file) {
            for line in lines {
                let record = line?;
                if let Ok(tx) = TxRecord::from_string_record(&record) {
                    let th = Self::hash(&tx);
                    // index for search the title and the tags
                    self.index.insert(
                        th,
                        &format!("{} {}", tx.get_name(), tx.get_tags().join(" ")),
                    );
                    // here is the move
                    self.data.insert(th, tx);
                }
            }
        }
        Ok(())
    }
    /// Persist the datastore to disk, overwriting existing files
    ///
    /// The order of the item saved is random
    pub fn save(&self, log_file: &Path) -> Result<(), std::io::Error> {
        let mut file = LineWriter::new(File::create(log_file)?);
        self.data.iter().for_each(|v| {
            file.write(v.1.to_string_record().as_bytes()).ok();
        });
        file.flush()?;
        Ok(())
    }
    /// Retrieve the cost of life for a date
    ///
    pub fn cost_of_life(&self, d: &NaiveDate) -> f32 {
        costoflife::cost_of_life(self.data.values(), d)
            .to_f32()
            .unwrap()
    }
    /// Perform a search for a string in tags and transaction name
    ///
    pub fn search(&self, pattern: &str) -> Vec<(String, f32, f32, String, String, f32, String)> {
        self.index
            .search(pattern)
            .iter()
            .map(|h| {
                let tx = self.data.get(h).unwrap();
                (
                    tx.get_name().to_string(),
                    tx.get_amount_total().to_f32().unwrap(),
                    tx.per_diem().to_f32().unwrap(),
                    tx.get_starts_on().to_string(),
                    tx.get_ends_on().to_string(),
                    tx.get_progress(None),
                    tx.get_tags().join("/"),
                )
            })
            .collect()
    }
    /// Compile a summary of the active costs, returning a tuple with
    /// (title, total amount, cost per day, percentage payed)
    pub fn summary(&self, d: &NaiveDate) -> Vec<(String, f32, f32, f32)> {
        let mut s = self
            .data
            .iter()
            .filter(|(_k, v)| v.is_active_on(d))
            .map(|(_k, v)| {
                (
                    String::from(v.get_name()),
                    v.get_amount_total().to_f32().unwrap(),
                    v.per_diem().to_f32().unwrap(),
                    v.get_progress(Some(*d)),
                )
            })
            .collect::<Vec<(String, f32, f32, f32)>>();
        // sort the results descending by completion
        s.sort_by(|a, b| (b.3).partial_cmp(&a.3).unwrap());
        s
    }
    /// Return aggregation summary for tags
    ///
    pub fn tags(&self, d: &NaiveDate) -> Vec<(String, usize, f32)> {
        // counters here
        let mut agg: HashMap<String, (usize, BigDecimal)> = HashMap::new();
        // aggregate tags
        self.data
            .iter()
            .filter(|(_h, tx)| tx.is_active_on(d))
            .for_each(|(_h, tx)| {
                tx.get_tags().iter().for_each(|tg| {
                    let (n, a) = match agg.get(tg) {
                        Some((n, a)) => (n + 1, a + tx.per_diem()),
                        None => (1, tx.per_diem()),
                    };
                    agg.insert(tg.to_string(), (n, a));
                    // * agg.entry(*tg).or_insert((1, tx.per_diem())) +=(1, tx.per_diem());
                });
            });
        // return
        let mut s = agg
            .iter()
            .map(|(tag, v)| (tag.to_string(), v.0, v.1.clone().to_f32().unwrap()))
            .collect::<Vec<(String, usize, f32)>>();
        // sort the results descending by count
        s.sort_by(|a, b| (b.2).partial_cmp(&a.2).unwrap());
        s
    }
    /// Insert a new tx record
    /// if the record exists returns the existing one
    ///
    /// TODO: handle duplicates more gracefully
    pub fn insert(&mut self, tx: &TxRecord) -> Option<TxRecord> {
        let th = Self::hash(tx);
        // index for search the title and the tags
        self.index.insert(
            th,
            &format!("{} {}", tx.get_name(), tx.get_tags().join(" ")),
        );
        self.data.insert(th, tx.clone())
    }
    /// Get the size of the datastore
    ///
    /// # Arguments
    ///
    /// * `on` - A Option<chrono:NaiveDate> to filter for active transactions
    ///
    /// if the Option is None then the full size is returned
    ///
    pub fn size(&self, on: Option<NaiveDate>) -> usize {
        match on {
            Some(date) => self.summary(&date).len(),
            None => self.data.len(),
        }
    }
    // The output is wrapped in a Result to allow matching on errors
    // Returns an Iterator to the Reader of the lines of the file.
    fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
    where
        P: AsRef<Path>,
    {
        let file = File::open(filename)?;
        Ok(io::BufReader::new(file).lines())
    }
    /// Compute the blake3 has for a TxRecord
    ///
    /// The hash is calculated on
    /// - name
    /// - lifetime
    /// - starts_on
    /// - amount
    ///
    fn hash(tx: &TxRecord) -> blake3::Hash {
        let fields = format!(
            "{}:{}:{}:{}",
            tx.get_name(),
            tx.get_amount(),
            tx.get_lifetime(),
            tx.get_starts_on(),
        );
        blake3::hash(fields.as_bytes())
    }
}
#[cfg(test)]
mod tests {
    use super::*;
    use ::costoflife::{self, TxRecord};
    #[test]
    fn test_datastore() {
        let mut ds = DataStore::new();
        // insert one entry
        ds.insert(&TxRecord::new("Test#1", "10").unwrap());
        ds.insert(&TxRecord::new("Test#2", "10").unwrap());
        // simple insert
        assert_eq!(ds.cost_of_life(&costoflife::today()), 20.0);
        // summary test
        let summary = ds.summary(&costoflife::today());
        assert_eq!(summary.len(), 2);
        // test tags
        let mut ds = DataStore::new();
        // insert one entry
        ds.insert(&TxRecord::from_str("Test#1 10€ #tag1").unwrap());
        ds.insert(&TxRecord::from_str("Test#2 20€ #tag2").unwrap());
        ds.insert(&TxRecord::from_str("Test#3 50€ #tag3").unwrap());
        ds.insert(&TxRecord::from_str("Test#4 40€ #tag2").unwrap());
        let tags = ds.tags(&costoflife::today());
        assert_eq!(tags.len(), 3);
        // tag2
        let got = &tags[0];
        let exp = (String::from("tag2"), 2 as usize, 60.0);
        assert_eq!(*got, exp);
        // tag3
        let got = &tags[1];
        let exp = (String::from("tag3"), 1 as usize, 50.0);
        assert_eq!(*got, exp);
        // test search
        assert_eq!(ds.search("tag").len(), 4);
        assert_eq!(ds.search("whatever").len(), 0);
        // test load
        let mut ds = DataStore::new();
        // db path
        let p = Path::new("./testdata/costoflife.data.txt");
        // load
        let r = ds.load(p);
        assert_eq!(r.is_err(), false);
        assert_eq!(ds.size(None), 5 as usize);
    }
}


src/interaction.rs
use ::costoflife::{parse_amount, today, CostOfLifeError, TxRecord};
use dialoguer::console::Term;
use dialoguer::{theme::ColorfulTheme, Confirm, Input, Password, Select};

use Feat::*;
use PolarAnswer::*;

#[derive(PartialEq)]
pub enum PolarAnswer {
    Yes,
    No,
}

impl PolarAnswer {
    pub fn to_bool(&self) -> bool {
        match self {
            Self::Yes => true,
            Self::No => false,
        }
    }

    pub fn from_bool(v: bool) -> PolarAnswer {
        match v {
            true => Self::Yes,
            false => Self::No,
        }
    }
}

#[derive(PartialEq)]
pub enum Feat {
    NonEmpty,
    Empty,
}

impl Feat {
    fn to_bool(&self) -> bool {
        match self {
            Self::NonEmpty => false,
            Self::Empty => true,
        }
    }
}

/// shortcut for Confirm
pub fn confirm(q: &str, def: PolarAnswer) -> PolarAnswer {
    PolarAnswer::from_bool(
        Confirm::with_theme(&ColorfulTheme::default())
            .with_prompt(q)
            .default(def.to_bool())
            .interact()
            .unwrap(),
    )
}

/// shortcut for Input
pub fn input(q: &str, empty: Feat) -> String {
    Input::with_theme(&ColorfulTheme::default())
        .with_prompt(q)
        .allow_empty(empty.to_bool())
        .interact()
        .unwrap()
}

/// shortcut for Select optional input
pub fn select_opt<'a, T: ?Sized>(q: &str, opts: Vec<(&'a str, &'a T)>) -> Option<&'a T> {
    match Select::with_theme(&ColorfulTheme::default())
        .with_prompt(q)
        .items(
            &opts
                .iter()
                .map(|(l, _v)| l.to_string())
                .collect::<Vec<String>>(),
        )
        .default(0)
        .interact_on_opt(&Term::stdout())
        .unwrap()
    {
        Some(i) => Some(opts[i].1),
        _ => None,
    }
}

/// Select an item
pub fn select<'a, T: ?Sized>(q: &str, opts: Vec<(&'a str, &'a T)>) -> &'a T {
    opts[Select::with_theme(&ColorfulTheme::default())
        .with_prompt(q)
        .items(
            &opts
                .iter()
                .map(|(l, _v)| l.to_string())
                .collect::<Vec<String>>(),
        )
        .default(0)
        .interact_on(&Term::stdout())
        .unwrap()]
    .1
}

/// Show the options
pub fn menu() -> Option<String> {
    // ask for the quality
    match select_opt(
        "hello there, what's up? esc/q to quit",
        vec![
            ("Summary", "summary"),
            ("Tags", "agenda"),
            ("New Tx", "today"),
        ],
    ) {
        Some(x) => Some(x.to_string()),
        _ => None,
    }
}

pub fn new_tx() -> Result<TxRecord, CostOfLifeError> {
    let name = input("What it is it about?", NonEmpty);
    // amount
    let mut amount = parse_amount("0.0").unwrap();
    loop {
        let v = input("how much does it cost?", NonEmpty);
        amount = match parse_amount(&v) {
            Some(a) => a,
            None => continue,
        }
    }
    // tags
    let mut tags: Vec<String> = Vec::new();
    while Yes == confirm("add a tag?", No) {
        tags.push(input("tag label: ", NonEmpty));
    }
    let starts_on = today();

    //
    //TxRecord::from(name, tags, amount, starts_on, Lifetime::SingleDay, )
    Err(CostOfLifeError::GenericError("Not implemented".to_owned()))
}


src/main.rs
mod ledger;
use ledger::DataStore;

mod interaction;

use std::fmt;

use clap::{Arg, Command};
use dialoguer::{theme::ColorfulTheme, Confirm};
use directories_next::ProjectDirs;
use pad::{Alignment, PadStr};

use std::error;
use std::fs;
use std::path::Path;
use std::str::FromStr;

use Alignment::*;
use Cell::*;

const VERSION: &str = env!("CARGO_PKG_VERSION");
const DB_FILENAME: &str = "costoflife.data.txt";

fn main() -> Result<(), Box<dyn error::Error>> {
    //println!("Welcome to CostOf.Life!");

    let matches = Command::new("costoflife")
        .version(VERSION)
        .author("Andrea G. <no.andrea@gmail.com>")
        .about("keep track of the cost of your daily life")
        .after_help("visit https://thecostof.life for more info")
        .arg(
            Arg::new("config")
                .short('c')
                .long("config")
                .value_name("FILE")
                .help("Sets a custom config file")
                .takes_value(true),
        )
        .arg(
            Arg::new("on_date")
                .short('o')
                .long("on")
                .value_name("DATE")
                .help("use this date to calculate the cost of life")
                .takes_value(true),
        )
        .subcommand(
            Command::new("add")
                .about("add new expense")
                .arg(
                    Arg::new("EXP_STR")
                        .help("write the expense string")
                        .required(true)
                        .multiple_occurrences(true)
                        .value_terminator("."),
                )
                .arg(
                    Arg::new("non_interactive")
                        .long("yes")
                        .short('y')
                        .takes_value(false)
                        .help("automatically reply yes"),
                ),
        )
        .subcommand(Command::new("summary").about("print th expenses summary"))
        .subcommand(Command::new("tags").about("print th expenses tags summary"))
        .subcommand(
            Command::new("search")
                .about("search for a transaction")
                .arg(
                    Arg::new("SEARCH_PATTERN")
                        .help("pattern to match for tags and/or tx name")
                        .required(true)
                        .multiple_occurrences(true)
                        .value_terminator("."),
                ),
        )
        .get_matches();

    // first, see if there is the config dir
    let path = match ProjectDirs::from("com", "FarcastTo", "CostOf.Life") {
        Some(p) => {
            if !p.data_dir().exists() {
                let authorized = Confirm::with_theme(&ColorfulTheme::default())
                    .with_prompt("The CostOf.Life data dir does not exists, can I create it?")
                    .default(true)
                    .interact()
                    .unwrap();
                if !authorized {
                    println!("nevermind then :(");
                    return Ok(());
                }
                match fs::create_dir_all(p.data_dir()) {
                    Ok(_) => println!("data folder created at {:?}", p.data_dir()),
                    Err(e) => {
                        println!("error creating folder {:?}: {}", p.data_dir(), e);
                        panic!()
                    }
                }
            }
            p.data_dir().join(Path::new(DB_FILENAME))
        }
        None => panic!("cannot retrieve the config file dir"),
    };
    // load the datastores
    let mut ds = DataStore::new();
    ds.load(path.as_path())?;
    // get the date
    let target_date = match matches.value_of("on_date") {
        Some(v) => costoflife::date_from_str(v).expect("The date provided is not valid"),
        None => costoflife::today(),
    };
    // command line
    match matches.subcommand() {
        Some(("add", c)) => {
            if let Some(values) = c.values_of("EXP_STR") {
                let v = values.collect::<Vec<&str>>().join(" ");
                let tx = costoflife::TxRecord::from_str(&v).expect("Cannot parse the input string");
                // check the values for
                if c.is_present("non_interactive") {
                    ds.insert(&tx);
                    ds.save(path.as_path())?;
                    println!("done!");
                    return Ok(());
                }
                // print the transaction
                println!("Name     : {}", tx.get_name());
                println!("Tags     : {}", tx.get_tags().join(", "));
                print!("Amount   : {}", tx.get_amount());
                if !tx.amount_is_total() {
                    print!("(Total: {}€)", tx.get_amount_total());
                }
                println!("\nFrom - To: {} - {}", tx.get_starts_on(), tx.get_ends_on());
                println!("Per Diem : {}", tx.per_diem());
                // save to the store
                match Confirm::with_theme(&ColorfulTheme::default())
                    .with_prompt("Do you want to add it?")
                    .default(true)
                    .interact()
                {
                    Ok(true) => {
                        ds.insert(&tx);
                        ds.save(path.as_path())?;
                        println!("done!")
                    }
                    _ => println!("ok, another time"),
                }
            } else {
                println!("Tell me what to add, eg: Car 2000€ .transport 5y")
            }
        }
        Some(("summary", _c)) => {
            let mut p = Printer::new(vec![27, 12, 9, 100]);
            // title
            p.head(vec!["Item", "Price", "Diem", "Progress"]);
            p.sep();

            // data
            ds.summary(&target_date)
                .iter()
                .for_each(|(itm, total, per_diem, prog)| {
                    // ⧚ ░ ◼ ▪ this are characters that can be used for the bar
                    p.row(vec![
                        Str(itm.to_string()),
                        Amt(*total),
                        Amt(*per_diem),
                        Pcent(*prog), // completion percentage
                    ]);
                });
            // separator
            p.sep();
            p.render();
        }
        Some(("tags", _c)) => {
            let mut p = Printer::new(vec![27, 12, 9, 100]);

            p.head(vec!["Title", "Count", "Diem", "%"]);
            p.sep();

            // total per diem
            let total = ds.cost_of_life(&target_date);
            // data
            ds.tags(&target_date).iter().for_each(|(tag, count, cost)| {
                p.row(vec![
                    Str(tag.to_string()),
                    Cnt(*count),
                    Amt(*cost),
                    Pcent(cost / total), // tag amount over total
                ]);
            });
            // separator
            p.sep();
            p.render();
        }
        Some(("search", c)) => {
            let mut p = Printer::new(vec![40, 12, 8, 11, 11, 30, 40]);

            if let Some(values) = c.values_of("SEARCH_PATTERN") {
                let pattern = values.collect::<Vec<&str>>().join(" ");
                // no results
                let res = ds.search(&pattern);
                if res.is_empty() {
                    println!("No matches found ¯\\_(ツ)_/¯");
                    return Ok(());
                }
                // with results
                p.head(vec!["Item", "Price", "Diem", "Start", "End", "Tags", "%"]);
                p.sep();
                // compute the total
                let mut totals = (0.0, 0.0);
                // data
                res.iter()
                    .for_each(|(itm, price, diem, s, e, pcent, tags)| {
                        p.row(vec![
                            Str(itm.to_string()),
                            Amt(*price),
                            Amt(*diem),
                            Str(s.to_string()),
                            Str(e.to_string()),
                            Str(tags.to_string()),
                            Pcent(*pcent),
                        ]);
                        totals = (totals.0 + price, totals.1 + diem);
                    });
                // separator
                p.sep();
                // print the total as well
                p.row(vec![
                    Empty,
                    Amt(totals.0),
                    Amt(totals.1),
                    Empty,
                    Empty,
                    Empty,
                    Empty,
                ]);
                p.render();
            }
        }
        Some((&_, _)) | None => {}
    }
    println!("Today CostOf.Life is: {}€", ds.cost_of_life(&target_date));
    Ok(())
}

#[derive(Debug)]
enum Cell {
    Amt(f32),    // amount
    Pcent(f32),  // percent
    Str(String), // string
    Cnt(usize),  // counter
    Empty,
    Sep,
}

#[derive(Debug)]
struct Printer {
    sizes: Vec<usize>,
    data: Vec<Vec<Cell>>,
    col_sep: String,
    row_sep: char,
    progress: char,
}

impl fmt::Display for Printer {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "{}",
            self.data
                .iter()
                .map(|row| {
                    row.iter()
                        .enumerate()
                        .map(|(i, c)| {
                            let s = self.sizes[i];
                            match c {
                                Str(v) => v.pad(s, ' ', Left, true),
                                Amt(v) => format!("{}€", v).pad(s, ' ', Right, false),
                                Cnt(v) => format!("{}", v).pad(s, ' ', Right, false),
                                Empty => "".pad(s, ' ', Right, false),
                                Pcent(v) => {
                                    let p = v * 100.0;
                                    let b = (p as usize * s) / 100; // bar length
                                    format!("{:.2}", p).pad(b, self.progress, Right, false)
                                }
                                Sep => "".pad(s, self.row_sep, Alignment::Right, false),
                            }
                        })
                        .collect::<Vec<String>>()
                        .join(&self.col_sep)
                })
                .collect::<Vec<String>>()
                .join("\n")
        )
    }
}

impl Printer {
    pub fn new(col_sizes: Vec<usize>) -> Printer {
        Printer {
            sizes: col_sizes,
            data: Vec::new(),
            row_sep: '-',
            progress: '▮',
            col_sep: "|".to_string(),
        }
    }

    pub fn row(&mut self, row_data: Vec<Cell>) {
        self.data.push(row_data);
    }

    pub fn head(&mut self, head_data: Vec<&str>) {
        self.row(head_data.iter().map(|v| Str(v.to_string())).collect());
    }

    pub fn sep(&mut self) {
        self.row(self.sizes.iter().map(|_| Sep).collect());
    }

    pub fn render(&self) {
        println!("{}", self);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_printer() {
        let mut p = Printer::new(vec![5, 10, 10, 50]);
        p.head(vec!["a", "b", "c", "d"]);
        p.sep();
        p.row(vec![
            Str("One".to_string()),
            Amt(80.0),
            Cnt(100),
            Pcent(0.1043), // completion percentage
        ]);
        p.row(vec![
            Str("Two".to_string()),
            Amt(59.0),
            Cnt(321),
            Pcent(0.0420123123), // completion percentage
        ]);
        p.row(vec![
            Str("Three".to_string()),
            Amt(220.0),
            Cnt(11),
            Pcent(0.309312321), // completion percentage
        ]);
        p.sep();

        let printed =
            "a    |b         |c         |d                                                 
-----|----------|----------|--------------------------------------------------
One  |       80€|       100|10.43
Two  |       59€|       321|4.20
Three|      220€|        11|▮▮▮▮▮▮▮▮▮▮30.93
-----|----------|----------|--------------------------------------------------";

        assert_eq!(p.data.len(), 6);
        assert_eq!(p.to_string(), printed);
    }
}
